stub/rawkv.go

```diff
@@ -0,0 +1,138 @@
+// limitations under the License.
+
+package stub
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// RawClientStub can be used like a rawkv.Client while it redirects all function
+// calls to an httpproxy server.
+type RawClientStub struct {
+	client      http.Client
+	proxyServer string
+	id          string
+}
+
+// NewRawClientStub creates a client for rawkv calls.
+func NewRawClientStub(proxyServer string, pdServers []string) (*RawClientStub, error) {
+	client := &RawClientStub{
+		client:      http.Client{Timeout: time.Second * 10},
+		proxyServer: strings.TrimSuffix(proxyServer, "/"),
+	}
+	res, err := client.send("/rawkv/client/new", &RawRequest{PDAddrs: pdServers})
+	if err != nil {
+		return nil, err
+	}
+	client.id = res.ID
+	return client, nil
+}
+
+// Close closes the client and releases resources in proxy server.
+func (c *RawClientStub) Close() error {
+	_, err := c.send(fmt.Sprintf("/rawkv/client/%s/close", c.id), &RawRequest{})
+	return err
+}
+
+// Get queries value with the key. When the key does not exist, it returns `nil, nil`.
+func (c *RawClientStub) Get(key []byte) ([]byte, error) {
+	res, err := c.send(fmt.Sprintf("/rawkv/client/%s/get", c.id), &RawRequest{Key: key})
+	if err != nil {
+		return nil, err
+	}
+	return res.Value, nil
+}
+
+// BatchGet queries values with the keys.
+func (c *RawClientStub) BatchGet(keys [][]byte) ([][]byte, error) {
+	res, err := c.send(fmt.Sprintf("/rawkv/client/%s/batch-get", c.id), &RawRequest{Keys: keys})
+	if err != nil {
+		return nil, err
+	}
+	return res.Values, nil
+}
+
+// Put stores a key-value pair to TiKV.
+func (c *RawClientStub) Put(key, value []byte) error {
+	_, err := c.send(fmt.Sprintf("/rawkv/client/%s/put", c.id), &RawRequest{Key: key, Value: value})
+	return err
+}
+
+// BatchPut stores key-value pairs to TiKV.
+func (c *RawClientStub) BatchPut(keys, values [][]byte) error {
+	_, err := c.send(fmt.Sprintf("/rawkv/client/%s/batch-put", c.id), &RawRequest{Keys: keys, Values: values})
+	return err
+}
+
+// Delete deletes a key-value pair from TiKV.
+func (c *RawClientStub) Delete(key []byte) error {
+	_, err := c.send(fmt.Sprintf("/rawkv/client/%s/delete", c.id), &RawRequest{Key: key})
+	return err
+}
+
+// BatchDelete deletes key-value pairs from TiKV.
+func (c *RawClientStub) BatchDelete(keys [][]byte) error {
+	_, err := c.send(fmt.Sprintf("/rawkv/client/%s/batch-delete", c.id), &RawRequest{Keys: keys})
+	return err
+}
+
+// DeleteRange deletes all key-value pairs in a range from TiKV.
+func (c *RawClientStub) DeleteRange(startKey, endKey []byte) error {
+	_, err := c.send(fmt.Sprintf("/rawkv/client/%s/delete-range", c.id), &RawRequest{StartKey: startKey, EndKey: endKey})
+	return err
+}
+
+// Scan queries continuous kv pairs in range [startKey, endKey), up to limit pairs.
+func (c *RawClientStub) Scan(startKey, endKey []byte, limit int) ([][]byte, [][]byte, error) {
```

Comment:
Do we also need `ReverseScan`?

Reply:
Yes. Plan to do it when resolving https://github.com/tikv/client-validator/issues/6